DigitalLearningSolutions.Web.Tests/Controllers/ResetPasswordControllerTests.cs

```diff
@@ -0,0 +1,108 @@
+        private ResetPasswordController unauthenticatedController;
+        private IPasswordResetService passwordResetService;
+
+        [SetUp]
+        public void SetUp()
+        {
+            passwordResetService = A.Fake<IPasswordResetService>();
+
+            var unauthenticatedUser = new ClaimsPrincipal(new ClaimsIdentity(string.Empty));
+            unauthenticatedController = GetControllerWithUser(unauthenticatedUser);
+
+            var authenticatedUser = new ClaimsPrincipal(new ClaimsIdentity("mock"));
+            authenticatedController = GetControllerWithUser(authenticatedUser);
+        }
+
+        [Test]
+        public async Task Index_should_redirect_to_homepage_if_user_is_authenticated()
+        {
+            // When
+            var result = await authenticatedController.Index("email", "code");
+
+            // Then
+            result.Should().BeRedirectToActionResult().WithControllerName("Home").WithActionName("Index");
+        }
+
+        [Test]
+        public async Task Index_should_render_if_user_is_unauthenticated_and_query_params_are_valid()
+        {
+            // Given
+            A.CallTo(() => passwordResetService.GetValidMatchingUserReferencesAsync("email", "code"))
+                .Returns(Task.FromResult(new List<UserReference> { new UserReference() }));
+
+            // When
+            var result = await unauthenticatedController.Index("email", "code");
+
+            // Then
+            result.Should().BeViewResult().WithDefaultViewName();
+        }
+
+        [Test]
+        public async Task Index_should_redirect_to_error_page_if_user_is_unauthenticated_and_query_params_are_invalid()
+        {
+            // Given
+            A.CallTo(() => passwordResetService.GetValidMatchingUserReferencesAsync("email", "code"))
+                .Returns(Task.FromResult(new List<UserReference>()));
+
+            // When
+            var result = await unauthenticatedController.Index("email", "code");
+
+            // Then
+            result.Should().BeRedirectToActionResult().WithActionName("Error");
+        }
+
+        [Test]
+        public async Task Index_should_redirect_to_login_page_if_user_is_unauthenticated_and_email_is_missing()
+        {
+            // When
+            var result = await unauthenticatedController.Index(null, "code");
+
+            // Then
+            result.Should().BeRedirectToActionResult().WithControllerName("Login").WithActionName("Index");
+        }
+
+        [Test]
+        public async Task Index_should_redirect_to_login_page_if_user_is_unauthenticated_and_code_is_missing()
+        {
+            // When
+            var result = await unauthenticatedController.Index("email", null);
+
+            // Then
+            result.Should().BeRedirectToActionResult().WithControllerName("Login").WithActionName("Index");
+        }
+
+        private ResetPasswordController GetControllerWithUser(ClaimsPrincipal user)

```

Comment:
Really nice tests! ⭐ 

One thing, Daniel set up a ControllerContextHelper which should allow you to write something like this:
```
new Controller()
                .WithDefaultContext()
                .WithMockUser(false);
```